File: backend/schemas/etf.py
from pydantic import BaseModel
from typing import Optional


class EtfInfo(BaseModel):
    ticker: str
    name: Optional[str] = None
    domicile: Optional[str] = None
    gross_expense_ratio: Optional[str] = None
    nav: Optional[str] = None
    aum: Optional[str] = None
    last_updated: Optional[str] = None


class EtfListResponse(BaseModel):
    etfs: list[EtfInfo]
    total: int


class EtfDatesResponse(BaseModel):
    ticker: str
    dates: list[str]

File: frontend/src/types/etf.ts
export interface EtfInfo {
  ticker: string
  name: string | null
  domicile: string | null
  gross_expense_ratio: string | null
  nav: string | null
  aum: string | null
  last_updated: string | null
}

export interface EtfListResponse {
  etfs: EtfInfo[]
  total: number
}

export interface HoldingItem {
  component_identifier: string | null
  component_name: string | null
  component_ticker: string | null
  component_weight: number | null
  component_sector: string | null
  component_shares: number | null
  component_currency: string | null
  component_sedol: string | null
}

export interface EtfDetailResponse extends EtfInfo {
  latest_composition_date: string | null
  available_dates: string[]
  holdings_count: number
  holdings: HoldingItem[]
  holdings_truncated: boolean
}

export interface CompositionResponse {
  ticker: string
  composition_date: string
  nav: number | null
  shares_outstanding: number | null
  total_net_assets: number | null
  holdings_count: number
  holdings: HoldingItem[]
  holdings_truncated: boolean
}
